### ISSUE TITLE
Kestrel port binding default to 80 or 443 / No error/warn/info/debug when "{abc}" provided

### ISSUE THREAD

**Issue by tebeco** [dotnet/aspnetcore#24384]
Hello 👋

## Describe the bug
Invalid `Http/Https` binding can be defaulted without any log and go on

After few long minutes (hours?) of debugging we found out that invalid port configuration will not say anything and silent.
Why did it takes so long ?
Because we were looking for the https binding that was `https://*:8443` on our POD, and it seems that another `appsettings...json` file still had an http://*:{place.holder} that was not replaced
So it seems that it defaulted to port 80 without saying anything at all.

The behavior was `AccessDeniedException` on a `Socket` binding attempt (because Pods should not run as root so any binding bellow 1024 is forbidden)

## Questions
so I wonder about 3 things now :
* should the SocketException / AccessDenied be explicit on the full binding url that was denied ? => http://[::]:80 or http://*:80`
* should there be at least a WARNING on the fact that a string with letter and placerhoder is probably not a valid URI / Port ?
* is there any form of URI that would makes this valid: http://*:{foo.bar.kix} with the port 80 ?

## To Reproduce
```pwsh
dotnet new webapi
dotnet publish
cd ./bin/Debug/net5.0/publish/
$env:ASPNETCORE_URLS="https://*:{abc}"
./foo.exe
```

![image](https://user-images.githubusercontent.com/2266487/88724385-a8db5700-d12a-11ea-81b5-7d94e8e4ac03.png)

### Further technical details
- ASP.NET Core version : both 3.1.x and 5.0-preview6
- Include the output of `dotnet --info`

**Comment by poke** [comment]
This appears to be a problem with [`BindingAdress.Parse`](https://github.com/dotnet/aspnetcore/blob/4aa5e03207a005d7310ba792c76f5339d74c3364/src/Http/Http/src/BindingAddress.cs#L78) which gets called early in the [`AddressBinder`](https://github.com/dotnet/aspnetcore/blob/4aa5e03207a005d7310ba792c76f5339d74c3364/src/Servers/Kestrel/Core/src/Internal/AddressBinder.cs#L104). Since the binder is not able to produce a better result from the input, it will eventually fall back to just binding on the port that the `BindingAdress` contains:

https://github.com/dotnet/aspnetcore/blob/4aa5e03207a005d7310ba792c76f5339d74c3364/src/Servers/Kestrel/Core/src/Internal/AddressBinder.cs#L135-L139

Unfortunately, `BindingAddress.Parse` does not appear to parse the input very well:

![BindingAddress.Parse producing a port 443](https://user-images.githubusercontent.com/132240/88723658-898ffa00-d129-11ea-94a1-688fc31fd8f3.png)

Calling `BindingAddress.Parse("https://*:{abc}")` will recognize `*:{abc}` as the host and assume [the default HTTPS port 443](https://github.com/dotnet/aspnetcore/blob/4aa5e03207a005d7310ba792c76f5339d74c3364/src/Http/Http/src/BindingAddress.cs#L147-L150).

**Comment by poke** [comment]
Digging a bit deeper, it seems that this part of `BindingAddress.Parse` is responsible:

https://github.com/dotnet/aspnetcore/blob/4aa5e03207a005d7310ba792c76f5339d74c3364/src/Http/Http/src/BindingAddress.cs#L131-L138

`portString` is correctly determined as `"{abc}"` but then there is an attempt to parse it which will obviously fail. Since there is no port, the default scheme port will be used and the `:{abc}` will be understood as part of the host.

To be honest, I don’t really understand why the logic was done like this. I would expect the parsing to fail when the identified portString does not end up being a number. After all, it existing means that there is a colon before, so the original string is `a:b` where `b` is not a number. I only see two options where a colon would be valid as part of the [authority](https://tools.ietf.org/html/rfc3986#section-3.2) of an URI though:
- As part of the user info, although this would require an `@` before the host name. And for the address binder, this wouldn’t really make any sense.
- Or for IPv6 addresses. Although then the part after the colon would be required to be a hexadecimal number, so curly braces definitely wouldn’t be allowed.

While supporting IPv6 addresses is obviously required in some way, I would expect the parsing to actually fail with some kind of error instead of blindly accepting the `:{abc}` here. Maybe we can fine tune this just a little bit so inputs like these would throw an error.

**Comment by BrennanConroy** [comment]
At a minimum we should log a warning.

**Comment by tebeco** [comment]
yeah as you said Warning would be the minimal ;)

I honestly think this should throw
especially since this is an explicit value being realllly wrong ^^

this is not "a defaulted" value

also i wonder if the exception should indicate the address tried to be bound

**Comment by Tratcher** [comment]
Hit this again. We should reconsider making this scenario throw, we're not binding to what the user said they wanted.

**Comment by tebeco** [comment]
yep, i guess anybody would expect to throw on that

i remember when opening this issue that i wondered if i could PR that during net5.
IIRC the code seemed rather specific than just a `int.TryParse` and i wasn't sure about the implications

**Comment by Tratcher** [comment]
Yeah, I think we'd add an `else` clause there that threw a more specific exception.

**Comment by h5aaimtron** [comment]
I've created a PR to address this issue. While investigating test failures with the addition of the else/throw code, it was determined that this block of code was being executed for named pipes. Since named pipes do not utilize ports, I modified the if check before the code block to only run when not a unix pipe and not a named pipe. This appears to have resolved the test errors.

**Comment by h5aaimtron** [comment]
@Tratcher @BrennanConroy I provided the PR with updated code. One of the biggest question marks regarding BindingAddress Parse method is whether it should validate the address/URI/pipe passed in and whether it should produce a valid bindable address. In its current form, it allows valid URIs that are not technically valid bindable addresses. If valid bindable addresses are not desired, Uri.TryCreate would probably be the better target for such work. If a valid bindable address, as inferred by the class name, is the goal, I believe my PR acts as a foundation for moving in that direction.
